### PR TITLE
(Wait for #2041) [SWAP] Implement lookahead(preloading) behavior

### PR DESCRIPTION
**Pull request by jihochu**
It implement lookahead(preloading) behavior.
Preloading is working with cache flush synchroneously, due to
cache flush is done when each execution order finishes.

Signed-off-by: Jiho Chu <jiho.chu@samsung.com>

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -122,6 +124,7 @@ nntrainer_conf.set('PLUGIN_INSTALL_PREFIX', nntrainer_libdir / 'nntrainer')
 nntrainer_conf.set('INCLUDE_INSTALL_DIR', nntrainer_includedir / '..')
 nntrainer_conf.set('MEMORY_SWAP', nntrainer_enable_swap)
 nntrainer_conf.set('MEMORY_SWAP_PATH', nntrainer_swapdir)
+nntrainer_conf.set('MEMORY_SWAP_LOOKAHEAD', nntrainer_swap_lookahead)
```
Comment (jijoongmoon) [inline meson.build:127]: I'm not sure lookahead is set as in nntrainer configuration. How about the model parameter?

File: nntrainer/models/neuralnet.cpp [hunk 2]
```diff
@@ -782,11 +784,11 @@ int NeuralNetwork::train_run(std::function<bool(void *userdata)> stop_cb) {
 
   auto train_for_iteration = [this, stop_cb](RunStats &stat,
                                              DataBuffer &buffer) {
-    model_graph.flushCache();
-
     forwarding(true, stop_cb);
     backwarding(iter++, stop_cb);
 
+    model_graph.flushCache();
```
Comment (jijoongmoon) [inline nntrainer/models/neuralnet.cpp:790]: Is there any reason to place this here? if it is, then please leave the comments.

File: nntrainer/tensor/manager.h [hunk 3]
```diff
@@ -30,6 +30,7 @@
 #include <vector>
 
 #include <basic_planner.h>
+#include <cache_loader.h>
```
Comment (jijoongmoon) [inline nntrainer/tensor/manager.h:33]: cache_loader.h is not included in this PR.

File: nntrainer/tensor/manager.cpp [hunk 4]
```diff
@@ -625,13 +625,51 @@ Manager::getWeights(const std::function<bool(const Weight *)> &condition) {
...
+  if (!swap_lookahead) {
+    weight_pool.flushCache();
+    tensor_pool.flushCache();
+  }
 }
 
 void Manager::flushCacheExcept(unsigned int order) {
-  weight_pool.flushCacheExcept(order);
-  tensor_pool.flushCacheExcept(order);
+  auto loadAsync = [&](TensorPool &pool, unsigned int order) {
```
Comment (jijoongmoon) [inline nntrainer/tensor/manager.cpp:635]: Not in this PR, but this ```flushCacheExecpt```is called before every execution order. forwarding, calcGradient, calcDerivatives. But for the backwarding, this is called in neuralnet.cpp, and for the forwarding, it is called in network_graph.cpp. So.. how about making forwrading_op like backwarding_op in nerualnet.cpp, and keeping the consistence?

File: nntrainer/tensor/task_executor.cpp [hunk 5]
```diff
@@ -26,78 +26,90 @@
...
+TaskExecutor::TaskExecutor(const std::string &n) :
+  name(n),
+  run_thread(true),
+  wait_complete(false) {
+  task_thread = std::thread([&]() {
+    ml_logd("Task Thread(%s): start thread", name.c_str());
+    while (run_thread) {
+      std::unique_lock lk(task_mutex);
+      if (!task_cv.wait_for(lk, std::chrono::milliseconds(500),
+                            [&] { return !task_queue.empty(); }))
```
Comment (jijoongmoon) [inline nntrainer/tensor/task_executor.cpp:40]: I suppose it takes quite a time to wait. there are many processes should be done before actual tasks are queued. such as parsing, making graph, compiling, memory planning, initializing, allocating, and loading training data..

File: nntrainer/tensor/manager.cpp [hunk 6]
```diff
@@ -625,13 +625,51 @@ Manager::getWeights(const std::function<bool(const Weight *)> &condition) {
...
+    auto t_fut = completed[std::get<1>(tasks)].get_future();
+    lock.unlock();
+
+    w_fut.wait();
+    t_fut.wait();
+
+    async_task_eos.erase(o);
+  };
+
+  // TODO: lookahead > 1 is required.
```
Comment (jijoongmoon) [inline nntrainer/tensor/manager.cpp:657]: I agree we need to do supporting when lookahead is greater than 1. and dynamic setting the Look a Head.

File: meson.build (hunk 1 above)
Comment (jihochu) [reply meson.build:127]: OK. it removed.

File: nntrainer/models/neuralnet.cpp (hunk 2 above)
Comment (jihochu) [reply nntrainer/models/neuralnet.cpp:790]: thanks. a comment is leaved.

File: nntrainer/tensor/manager.h (hunk 3 above)
Comment (jihochu) [reply nntrainer/tensor/manager.h:33]: thanks. it removed.

File: nntrainer/tensor/manager.cpp (hunk 4 above)
Comment (jihochu) [reply nntrainer/tensor/manager.cpp:635]: Good point. I'll modify for forwarding_op in next PR.

File: nntrainer/tensor/task_executor.cpp (hunk 5 above)
Comment (jihochu) [reply nntrainer/tensor/task_executor.cpp:40]: I think it's not big burden for cheching this, but we can adjust start time of thread.

File: nntrainer/tensor/manager.cpp (hunk 6 above)
Comment (jihochu) [reply nntrainer/tensor/manager.cpp:657]: OK^^

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Comment by SeoHyungjun** [comment]
LGTM!

**Review by baek2sm — APPROVED** [review]
LGTM

**Pull request merged**